File: clj/src/day06/day06.clj
(require '[day06.guard-map :as gm])

(def test-file "data/day06/input.txt")

(def guard-map
  (gm/parse-guard-map-file {:file-name test-file}))

(doall
 (for [line (:map (gm/visited-map {:guard-map guard-map}))]
   (println line)))

(def guard-path
  (gm/guard-path {:guard-map guard-map}))

(def visited-locations
  (->> guard-path
       (map (juxt :row :col))
       set))

(println {:visited-locations (count visited-locations)})

(def possible-loop-obstructions
  (gm/possible-loop-obstructions {:guard-map guard-map
                                  :guard-path guard-path}))

(println {:possible-loop-obstructions (count possible-loop-obstructions)})

File: day06/day06.zig
const std = @import("std");
const gm = @import("guard_map.zig");

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const guard_map = try gm.parseGuardMapFile(allocator, "data/day06/input.txt");
    const visited_map = try gm.visitGuardMap(allocator, guard_map);
    for (visited_map.map) |line| {
        std.debug.print("{s}\n", .{line});
    }
    std.debug.print("Guard visited {} positions\n", .{gm.countVisitedPositions(visited_map)});
}

test {
    _ = std.testing.refAllDecls(@This());
}
